File: optimization/framework/scorecard_focus.py
"""Convert deterministic scorecards into optimization focus hints.

This module intentionally lives under `optimization/`: evaluation produces
scorecards, and local optimization may consume them. The dependency direction
does not go the other way.
"""

from __future__ import annotations

from collections import Counter, defaultdict
import json
from typing import Any


def _failed_checks(case: dict[str, Any]) -> list[dict[str, Any]]:
    return [check for check in case.get("checks", []) if check.get("result") == "fail"]


def _visual_failure_detail(visual_review: dict[str, Any]) -> str:
    summary = str(visual_review.get("summary") or "").strip()
    dimensions = visual_review.get("dimensions", {})
    failing_dimensions = []
    if isinstance(dimensions, dict):
        for name, dimension in sorted(dimensions.items()):
            if not isinstance(dimension, dict):
                continue
            status = str(dimension.get("status") or "")
            score = dimension.get("score")
            if status in {"fail", "needs_review"} or (isinstance(score, (int, float)) and score < 6):
                failing_dimensions.append(name)
    if failing_dimensions:
        suffix = "Failing visual dimensions: " + ", ".join(failing_dimensions)
        return f"{summary} {suffix}".strip()
    return summary or "Visual review did not pass."


def _visual_failed_checks(case: dict[str, Any]) -> list[dict[str, Any]]:
    visual_review = case.get("visual_review")
    if not isinstance(visual_review, dict):
        return []
    status = str(visual_review.get("status") or "not_reviewed")
    if status in {"pass", "not_required"}:
        return []
    if (
        not visual_review.get("required")
        and not visual_review.get("blocking")
        and str(visual_review.get("reviewer") or "unassigned") == "unassigned"
    ):
        return []
    return [
        {
            "check_id": f"visual_review_{status}",
            "type": "qualitative_visual_review",
            "category": "visual_review",
            "critical": bool(visual_review.get("blocking", True)),
            "actual": status,
            "expected": "pass",
            "tolerance": None,
            "detail": _visual_failure_detail(visual_review),
        }
    ]


def _all_failed_checks(case: dict[str, Any]) -> list[dict[str, Any]]:
    return [*_failed_checks(case), *_visual_failed_checks(case)]


def _category_failures(scorecard: dict[str, Any]) -> Counter[str]:
    counts: Counter[str] = Counter()
    for case in scorecard.get("cases", []):
        for check in _all_failed_checks(case):
            counts[str(check.get("category", "uncategorized"))] += 1
    return counts


def _critical_counts(scorecard: dict[str, Any]) -> Counter[str]:
    counts: Counter[str] = Counter()
    for failure in scorecard.get("critical_failures", []):
        counts[str(failure.get("category", "uncategorized"))] += 1
    for case in scorecard.get("cases", []):
        for check in _visual_failed_checks(case):
            if check.get("critical"):
                counts[str(check.get("category", "uncategorized"))] += 1
    return counts


def _visual_score(scorecard: dict[str, Any]) -> float:
    summary = scorecard.get("visual_summary", {})
    if not isinstance(summary, dict):
        return 0.0
    required = int(summary.get("required_count") or summary.get("total_cases") or 0)
    if required <= 0:
        return 1.0
    passed = int(summary.get("pass_count") or 0)
    return passed / required


def build_focus(scorecard: dict[str, Any]) -> dict[str, Any]:
    threshold = float(scorecard.get("threshold", 0.95))
    category_failures = _category_failures(scorecard)
    critical_counts = _critical_counts(scorecard)
    category_scores = scorecard.get("category_scores", {})

    categories = []
    category_names = set(category_scores) | set(category_failures) | set(critical_counts)
    for category in sorted(category_names):
        data = category_scores.get(category, {})
        score = _visual_score(scorecard) if category == "visual_review" and category not in category_scores else float(data.get("score", 0.0))
        failed_checks = category_failures[category]
        critical_failures = critical_counts[category]
        if score >= threshold and failed_checks == 0 and critical_failures == 0:
            continue
        categories.append(
            {
                "category": category,
                "score": score,
                "threshold": threshold,
                "failed_checks": failed_checks,
                "critical_failures": critical_failures,
                "priority": critical_failures * 100 + failed_checks * 10 + max(0.0, threshold - score),
            }
        )
    categories.sort(key=lambda item: (-item["priority"], item["category"]))

    cases = []
    skills: Counter[str] = Counter()
    for case in scorecard.get("cases", []):
        failed = _all_failed_checks(case)
        if not failed:
            continue
        skill = str(case.get("skill", ""))
        skills[skill] += len(failed)
        cases.append(
            {
                "skill": skill,
                "case_id": case.get("case_id", ""),
                "case_name": case.get("case_name", ""),
                "task": case.get("task", ""),
                "evidence_path": case.get("evidence_path", ""),
                "score": case.get("score", 0.0),
                "failed_checks": [
                    {
                        "check_id": check.get("check_id", ""),
                        "type": check.get("type", ""),
                        "category": check.get("category", "uncategorized"),
                        "critical": bool(check.get("critical", False)),
                        "actual": check.get("actual"),
                        "expected": check.get("expected"),
                        "tolerance": check.get("tolerance"),
                        "detail": check.get("detail", ""),
                    }
                    for check in failed
                ],
            }
        )
    cases.sort(key=lambda item: (-sum(1 for check in item["failed_checks"] if check["critical"]), item["skill"], item["case_id"]))

    category_to_cases: dict[str, set[str]] = defaultdict(set)
    for case in cases:
        for check in case["failed_checks"]:
            category_to_cases[check["category"]].add(f"{case['skill']}/{case['case_id']}")

    return {
        "schema_version": "1.0",
        "source_run_id": scorecard.get("run_id", ""),
        "source_git_commit": scorecard.get("git_commit", ""),
        "source_result": scorecard.get("overall_result", ""),
        "source_score": scorecard.get("overall_score", 0.0),
        "threshold": threshold,
        "focus_required": bool(categories or cases),
        "categories": [
            {
                **category,
                "affected_cases": sorted(category_to_cases.get(category["category"], set())),
            }
            for category in categories
        ],
        "skills": [
            {"skill": skill, "failed_checks": failed_checks}
            for skill, failed_checks in sorted(skills.items(), key=lambda item: (-item[1], item[0]))
        ],
        "cases": cases,
    }


def _markdown_value(value: Any, max_length: int = 140) -> str:
    if value is None:
        text = "null"
    elif isinstance(value, str):
        text = value
    else:
        text = json.dumps(value, sort_keys=True)
    text = text.replace("\n", " ").replace("|", "\\|")
    return text if len(text) <= max_length else text[: max_length - 1] + "..."


def focus_to_markdown(focus: dict[str, Any]) -> str:
    lines = [
        "# Optimization Focus From Scorecard",
        "",
        f"- Source run: `{focus['source_run_id']}`",
        f"- Source result: **{str(focus['source_result']).upper()}**",
        f"- Source score: {float(focus['source_score']):.1%}",
        f"- Threshold: {float(focus['threshold']):.1%}",
        "",
    ]
    if not focus["focus_required"]:
        lines.append("No optimization focus is required from this scorecard.")
        return "\n".join(lines).rstrip() + "\n"

    lines.extend(["## Categories", "", "| Category | Score | Failed Checks | Critical Failures | Affected Cases |"])
    lines.append("| --- | ---: | ---: | ---: | --- |")
    for category in focus["categories"]:
        lines.append(
            f"| {category['category']} | {category['score']:.1%} | "
            f"{category['failed_checks']} | {category['critical_failures']} | "
            f"{', '.join(category['affected_cases'])} |"
        )

    lines.extend(["", "## Cases", ""])
    for case in focus["cases"]:
        lines.append(f"### {case['skill']} / {case['case_id']} - {case['case_name']}")
        lines.append(f"- Score: {float(case['score']):.1%}")
        lines.append(f"- Evidence: `{case['evidence_path']}`")
        for check in case["failed_checks"]:
            lines.append(
                f"- `{check['check_id']}` ({check['category']}): {check['detail']}; "
                f"actual={_markdown_value(check['actual'])}; "
                f"expected={_markdown_value(check['expected'])}; "
                f"tolerance={_markdown_value(check['tolerance'])}"
            )
        lines.append("")

    return "\n".join(lines).rstrip() + "\n"


def focus_to_decision(focus: dict[str, Any], *, skill: str | None = None) -> dict[str, Any]:
    """Create a proposer-compatible decision record from scorecard focus.

    The legacy proposer already knows how to consume a decision record as its
    strongest recent signal. This keeps the bridge local to optimization while
    allowing deterministic scorecard failures to seed a targeted proposal pass.
    """
    cases = [
        case
        for case in focus.get("cases", [])
        if skill is None or case.get("skill") == skill
    ]
    categories = []
    for category in focus.get("categories", []):
        affected = [
            case_id
            for case_id in category.get("affected_cases", [])
            if skill is None or str(case_id).startswith(f"{skill}/")
        ]
        if affected or skill is None:
            categories.append({**category, "affected_cases": affected})

    failed_checks = sum(len(case.get("failed_checks", [])) for case in cases)
    critical_failures = sum(
        1
        for case in cases
        for check in case.get("failed_checks", [])
        if check.get("critical")
    )
    category_summary = ", ".join(
        f"{category['category']} {float(category['score']):.1%}"
        for category in categories[:5]
    ) or "no failing categories"
    case_lines = []
    for case in cases[:8]:
        check_summaries = "; ".join(
            f"{check['check_id']}: {check['detail']}"
            for check in case.get("failed_checks", [])[:3]
        )
        case_lines.append(
            f"- {case['skill']}/{case['case_id']} {case['case_name']}: {check_summaries}"
        )
    rationale = (
        "Deterministic scorecard focus should guide the next local optimization. "
        f"Source result={focus.get('source_result')} score={float(focus.get('source_score', 0.0)):.1%} "
        f"threshold={float(focus.get('threshold', 0.95)):.1%}. "
        f"Failing categories: {category_summary}. "
        "Failed checks:\n" + ("\n".join(case_lines) if case_lines else "- none")
    )

    return {
        "decision": "SCORECARD_FOCUS" if cases else "SCORECARD_CLEAN",
        "rule_fired": "scorecard_critical_failure_focus" if critical_failures else "scorecard_threshold_focus",
        "rationale": rationale,
        "counts": {
            "wins": 0,
            "losses": failed_checks,
            "ties": 0,
        },
        "scorecard_focus": {
            "source_run_id": focus.get("source_run_id", ""),
            "source_git_commit": focus.get("source_git_commit", ""),
            "source_result": focus.get("source_result", ""),
            "source_score": focus.get("source_score", 0.0),
            "threshold": focus.get("threshold", 0.95),
            "skill": skill,
            "failed_checks": failed_checks,
            "critical_failures": critical_failures,
            "categories": categories,
            "cases": cases,
        },
    }

File: packages/eval/src/optimization/scorecardFocus.ts
/**
 * Convert deterministic scorecards into optimization focus hints.
 * Port of optimization/framework/scorecard_focus.py.
 */
import { pct } from "../lib/format.js";

function failedChecks(caseRow: Record<string, any>): Array<Record<string, any>> {
  return (caseRow.checks ?? []).filter((check: Record<string, any>) => check.result === "fail");
}

function visualFailureDetail(visualReview: Record<string, any>): string {
  const summary = String(visualReview.summary ?? "").trim();
  const dimensions = visualReview.dimensions ?? {};
  const failingDimensions: string[] = [];
  if (dimensions !== null && typeof dimensions === "object" && !Array.isArray(dimensions)) {
    for (const name of Object.keys(dimensions).sort()) {
      const dimension = dimensions[name];
      if (dimension === null || typeof dimension !== "object" || Array.isArray(dimension)) continue;
      const status = String(dimension.status ?? "");
      const score = dimension.score;
      if (["fail", "needs_review"].includes(status) || (typeof score === "number" && score < 6)) {
        failingDimensions.push(name);
      }
    }
  }
  if (failingDimensions.length) {
    const suffix = "Failing visual dimensions: " + failingDimensions.join(", ");
    return `${summary} ${suffix}`.trim();
  }
  return summary || "Visual review did not pass.";
}

function visualFailedChecks(caseRow: Record<string, any>): Array<Record<string, any>> {
  const visualReview = caseRow.visual_review;
  if (visualReview === null || visualReview === undefined || typeof visualReview !== "object" || Array.isArray(visualReview)) {
    return [];
  }
  const status = String(visualReview.status ?? "not_reviewed");
  if (["pass", "not_required"].includes(status)) return [];
  if (
    !visualReview.required &&
    !visualReview.blocking &&
    String(visualReview.reviewer ?? "unassigned") === "unassigned"
  ) {
    return [];
  }
  return [
    {
      check_id: `visual_review_${status}`,
      type: "qualitative_visual_review",
      category: "visual_review",
      critical: Boolean(visualReview.blocking ?? true),
      actual: status,
      expected: "pass",
      tolerance: null,
      detail: visualFailureDetail(visualReview),
    },
  ];
}

function allFailedChecks(caseRow: Record<string, any>): Array<Record<string, any>> {
  return [...failedChecks(caseRow), ...visualFailedChecks(caseRow)];
}

function categoryFailures(scorecard: Record<string, any>): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const caseRow of scorecard.cases ?? []) {
    for (const check of allFailedChecks(caseRow)) {
      const category = String(check.category ?? "uncategorized");
      counts[category] = (counts[category] ?? 0) + 1;
    }
  }
  return counts;
}

function criticalCounts(scorecard: Record<string, any>): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const failure of scorecard.critical_failures ?? []) {
    const category = String(failure.category ?? "uncategorized");
    counts[category] = (counts[category] ?? 0) + 1;
  }
  for (const caseRow of scorecard.cases ?? []) {
    for (const check of visualFailedChecks(caseRow)) {
      if (check.critical) {
        const category = String(check.category ?? "uncategorized");
        counts[category] = (counts[category] ?? 0) + 1;
      }
    }
  }
  return counts;
}

function visualScore(scorecard: Record<string, any>): number {
  const summary = scorecard.visual_summary ?? {};
  if (summary === null || typeof summary !== "object" || Array.isArray(summary)) return 0;
  const required = Number(summary.required_count ?? summary.total_cases ?? 0);
  if (required <= 0) return 1;
  const passed = Number(summary.pass_count ?? 0);
  return passed / required;
}

export function buildFocus(scorecard: Record<string, any>): Record<string, any> {
  const threshold = Number(scorecard.threshold ?? 0.95);
  const categoryFails = categoryFailures(scorecard);
  const criticals = criticalCounts(scorecard);
  const categoryScores = scorecard.category_scores ?? {};

  const categories: Array<Record<string, any>> = [];
  const categoryNames = new Set([...Object.keys(categoryScores), ...Object.keys(categoryFails), ...Object.keys(criticals)]);
  for (const category of [...categoryNames].sort()) {
    const data = categoryScores[category] ?? {};
    const score =
      category === "visual_review" && !(category in categoryScores)
        ? visualScore(scorecard)
        : Number(data.score ?? 0);
    const failedCount = categoryFails[category] ?? 0;
    const criticalCount = criticals[category] ?? 0;
    if (score >= threshold && failedCount === 0 && criticalCount === 0) continue;
    categories.push({
      category,
      score,
      threshold,
      failed_checks: failedCount,
      critical_failures: criticalCount,
      priority: criticalCount * 100 + failedCount * 10 + Math.max(0, threshold - score),
    });
  }
  categories.sort((a, b) => (b.priority - a.priority) || (a.category < b.category ? -1 : a.category > b.category ? 1 : 0));

  const cases: Array<Record<string, any>> = [];
  const skills: Record<string, number> = {};
  for (const caseRow of scorecard.cases ?? []) {
    const failed = allFailedChecks(caseRow);
    if (!failed.length) continue;
    const skill = String(caseRow.skill ?? "");
    skills[skill] = (skills[skill] ?? 0) + failed.length;
    cases.push({
      skill,
      case_id: caseRow.case_id ?? "",
      case_name: caseRow.case_name ?? "",
      task: caseRow.task ?? "",
      evidence_path: caseRow.evidence_path ?? "",
      score: caseRow.score ?? 0,
      failed_checks: failed.map((check) => ({
        check_id: check.check_id ?? "",
        type: check.type ?? "",
        category: check.category ?? "uncategorized",
        critical: Boolean(check.critical ?? false),
        actual: check.actual ?? null,
        expected: check.expected ?? null,
        tolerance: check.tolerance ?? null,
        detail: check.detail ?? "",
      })),
    });
  }
  cases.sort((a, b) => {
    const critA = a.failed_checks.filter((check: Record<string, any>) => check.critical).length;
    const critB = b.failed_checks.filter((check: Record<string, any>) => check.critical).length;
    if (critB !== critA) return critB - critA;
    if (a.skill !== b.skill) return a.skill < b.skill ? -1 : 1;
    return a.case_id < b.case_id ? -1 : a.case_id > b.case_id ? 1 : 0;
  });

  const categoryToCases: Record<string, Set<string>> = {};
  for (const caseRow of cases) {
    for (const check of caseRow.failed_checks) {
      (categoryToCases[check.category] ??= new Set()).add(`${caseRow.skill}/${caseRow.case_id}`);
    }
  }

  return {
    schema_version: "1.0",
    source_run_id: scorecard.run_id ?? "",
    source_git_commit: scorecard.git_commit ?? "",
    source_result: scorecard.overall_result ?? "",
    source_score: scorecard.overall_score ?? 0,
    threshold,
    focus_required: Boolean(categories.length || cases.length),
    categories: categories.map((category) => ({
      ...category,
      affected_cases: [...(categoryToCases[category.category] ?? new Set())].sort(),
    })),
    skills: Object.entries(skills)
      .sort(([skillA, countA], [skillB, countB]) => (countB - countA) || (skillA < skillB ? -1 : 1))
      .map(([skill, failedCount]) => ({ skill, failed_checks: failedCount })),
    cases,
  };
}

function markdownValue(value: unknown, maxLength = 140): string {
  let text: string;
  if (value === null || value === undefined) text = "null";
  else if (typeof value === "string") text = value;
  else {
    const sortDeep = (v: any): any => {
      if (Array.isArray(v)) return v.map(sortDeep);
      if (v !== null && typeof v === "object") {
        const out: Record<string, any> = {};
        for (const key of Object.keys(v).sort()) out[key] = sortDeep(v[key]);
        return out;
      }
      return v;
    };
    text = JSON.stringify(sortDeep(value));
  }
  text = text.replace(/\n/g, " ").replace(/\|/g, "\\|");
  return text.length <= maxLength ? text : text.slice(0, maxLength - 1) + "...";
}

export function focusToMarkdown(focus: Record<string, any>): string {
  const lines = [
    "# Optimization Focus From Scorecard",
    "",
    `- Source run: \`${focus.source_run_id}\``,
    `- Source result: **${String(focus.source_result).toUpperCase()}**`,
    `- Source score: ${pct(Number(focus.source_score))}`,
    `- Threshold: ${pct(Number(focus.threshold))}`,
    "",
  ];
  if (!focus.focus_required) {
    lines.push("No optimization focus is required from this scorecard.");
    return lines.join("\n").replace(/\s+$/, "") + "\n";
  }

  lines.push("## Categories", "", "| Category | Score | Failed Checks | Critical Failures | Affected Cases |");
  lines.push("| --- | ---: | ---: | ---: | --- |");
  for (const category of focus.categories) {
    lines.push(
      `| ${category.category} | ${pct(category.score)} | ` +
        `${category.failed_checks} | ${category.critical_failures} | ` +
        `${category.affected_cases.join(", ")} |`,
    );
  }

  lines.push("", "## Cases", "");
  for (const caseRow of focus.cases) {
    lines.push(`### ${caseRow.skill} / ${caseRow.case_id} - ${caseRow.case_name}`);
    lines.push(`- Score: ${pct(Number(caseRow.score))}`);
    lines.push(`- Evidence: \`${caseRow.evidence_path}\``);
    for (const check of caseRow.failed_checks) {
      lines.push(
        `- \`${check.check_id}\` (${check.category}): ${check.detail}; ` +
          `actual=${markdownValue(check.actual)}; ` +
          `expected=${markdownValue(check.expected)}; ` +
          `tolerance=${markdownValue(check.tolerance)}`,
      );
    }
    lines.push("");
  }

  return lines.join("\n").replace(/\s+$/, "") + "\n";
}

/** Create a proposer-compatible decision record from scorecard focus. */
export function focusToDecision(focus: Record<string, any>, skill: string | null = null): Record<string, any> {
  const cases = (focus.cases ?? []).filter((caseRow: Record<string, any>) => skill === null || caseRow.skill === skill);
  const categories: Array<Record<string, any>> = [];
  for (const category of focus.categories ?? []) {
    const affected = (category.affected_cases ?? []).filter(
      (caseId: string) => skill === null || String(caseId).startsWith(`${skill}/`),
    );
    if (affected.length || skill === null) {
      categories.push({ ...category, affected_cases: affected });
    }
  }

  const failedChecksCount = cases.reduce(
    (sum: number, caseRow: Record<string, any>) => sum + (caseRow.failed_checks ?? []).length,
    0,
  );
  const criticalFailures = cases.reduce(
    (sum: number, caseRow: Record<string, any>) =>
      sum + (caseRow.failed_checks ?? []).filter((check: Record<string, any>) => check.critical).length,
    0,
  );
  const categorySummary =
    categories
      .slice(0, 5)
      .map((category) => `${category.category} ${pct(Number(category.score))}`)
      .join(", ") || "no failing categories";
  const caseLines: string[] = [];
  for (const caseRow of cases.slice(0, 8)) {
    const checkSummaries = (caseRow.failed_checks ?? [])
      .slice(0, 3)
      .map((check: Record<string, any>) => `${check.check_id}: ${check.detail}`)
      .join("; ");
    caseLines.push(`- ${caseRow.skill}/${caseRow.case_id} ${caseRow.case_name}: ${checkSummaries}`);
  }
  const rationale =
    "Deterministic scorecard focus should guide the next local optimization. " +
    `Source result=${focus.source_result} score=${pct(Number(focus.source_score ?? 0))} ` +
    `threshold=${pct(Number(focus.threshold ?? 0.95))}. ` +
    `Failing categories: ${categorySummary}. ` +
    "Failed checks:\n" +
    (caseLines.length ? caseLines.join("\n") : "- none");

  return {
    decision: cases.length ? "SCORECARD_FOCUS" : "SCORECARD_CLEAN",
    rule_fired: criticalFailures ? "scorecard_critical_failure_focus" : "scorecard_threshold_focus",
    rationale,
    counts: { wins: 0, losses: failedChecksCount, ties: 0 },
    scorecard_focus: {
      source_run_id: focus.source_run_id ?? "",
      source_git_commit: focus.source_git_commit ?? "",
      source_result: focus.source_result ?? "",
      source_score: focus.source_score ?? 0,
      threshold: focus.threshold ?? 0.95,
      skill,
      failed_checks: failedChecksCount,
      critical_failures: criticalFailures,
      categories,
      cases,
    },
  };
}
